lib/uuid/uuid.c

```diff
@@ -0,0 +1,276 @@
+	return 0;
+}
+#endif
+
+#if defined(CONFIG_UUID_V5)
+int uuid_generate_v5(const uuid_t namespace, const void *data, size_t data_size, uuid_t out)
+{
+	if ((out == NULL) || (namespace == NULL)) {
+		return -EINVAL;
+	}
+	int mbedtls_err = 0;
+	mbedtls_md_context_t ctx = {0};
+	const mbedtls_md_info_t *md_info = mbedtls_md_info_from_type(MBEDTLS_MD_SHA1);
+	const size_t sha_1_bytes = 20;
+	uint8_t sha_result[sha_1_bytes];
+
+	mbedtls_md_init(&ctx);
+	mbedtls_err = mbedtls_md_setup(&ctx, md_info, 0);
+	/* Might return: MBEDTLS_ERR_MD_BAD_INPUT_DATA or MBEDTLS_ERR_MD_ALLOC_FAILED */
+	switch (mbedtls_err) {
+	case 0:
+		break;
+	case MBEDTLS_ERR_MD_BAD_INPUT_DATA:
+		return -EINVAL;
+	case MBEDTLS_ERR_MD_ALLOC_FAILED:
+		return -ENOMEM;
+	default:
+		return -ENOTSUP;
+	}
+	mbedtls_err = mbedtls_md_starts(&ctx);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_update(&ctx, namespace, UUID_SIZE);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_update(&ctx, data, data_size);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_err = mbedtls_md_finish(&ctx, sha_result);
+	if (mbedtls_err != 0) {
+		/* Might return MBEDTLS_ERR_MD_BAD_INPUT_DATA */
+		return -EINVAL;
+	}
+	mbedtls_md_free(&ctx);
+
+	/* Store the computed SHA1 in the out struct */
+	for (unsigned int i = 0; i < UUID_SIZE; i++) {
+		out[i] = sha_result[i];
+	}
+	/* Update version and variant */
+	overwrite_uuid_version_and_variant(out, UUID_V5_VERSION, UUID_V5_VARIANT);
+
+	return 0;
+}
+#endif
+
+int uuid_from_buffer_be(const uint8_t data[UUID_SIZE], uuid_t out)
+{
+	if ((data == NULL) || (out == NULL)) {
+		return -EINVAL;
+	}
+	memcpy(out, data, UUID_SIZE);
+	return 0;
+}
+
+int uuid_from_buffer_le(const uint8_t data[UUID_SIZE], uuid_t out)
+{
+	if ((data == NULL) || (out == NULL)) {
+		return -EINVAL;
+	}
+
+	const size_t first_part_start = 0U;
+	const size_t first_part_size = 4U;
+	const size_t second_part_start = first_part_start + first_part_size;
+	const size_t second_part_size = 2U;
+	const size_t third_part_start = second_part_start + second_part_size;
+	const size_t third_part_size = 2U;
+	const size_t fourth_part_start = third_part_start + third_part_size;
+	const size_t fourth_part_size = 2U;
+	const size_t fifth_part_start = fourth_part_start + fourth_part_size;
+	const size_t fifth_part_size = 6U;
+
+	memcpy_swap(&out[first_part_start], &data[first_part_start], first_part_size);
+	memcpy_swap(&out[second_part_start], &data[second_part_start], second_part_size);
+	memcpy_swap(&out[third_part_start], &data[third_part_start], third_part_size);
+	memcpy_swap(&out[fourth_part_start], &data[fourth_part_start], fourth_part_size);
+	memcpy(&out[fifth_part_start], &data[fifth_part_start], fifth_part_size);
```

Comment:
In the documentation for this function, only the first 3 parts are swapped, but this swaps the first 4. Why?

And why isn't the 5th swapped? (not familiar with the UUID spec)

Reply:
The issue with the endianness is that the UUID spec defines UUID only as big-endian.
The interpretation of little-endian UUID is not universal.
From my understanding here are our options for the UUID:
- Standard big-endian representation: `00112233-4455-6677-8899-AABBCCDDEEFF`
- little-endian representation inverting each block: `33221100-5544-7766-9988-FFEEDDCCBBAA`
- little-endian representation inverting the full UUID: `FFEEDDCC-BBAA-9988-7766-554433221100`
- little-endian representation for Microsoft COM GUIs representation: `33221100-5544-7766-8899-AABBCCDDEEFF`

I'm not sure which of those little-endian notations is used in the Bluetooth module. 
However, from my very superficial research, I think the last one is the more widely adopted.

Reply:
I'll need to read up on the UUID spec. Endianess is always a bit tricky, but it usually comes down the the language and terminology used in the specs and whether the blocks are values or arrays of values.

Reply:
The blocks change based on the UUID version. For example, v5 is defined as follows: 
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           sha1_high                           |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|         sha1_high             |  ver  |      sha1_mid         |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|var|                       sha1_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           sha1_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
Where sha1_* are all components of the same SHA-1 hash.
While UUID v1 is defined as follows:
```
 0                   1                   2                   3
 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                           time_low                            |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|           time_mid            |  ver  |       time_high       |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|var|         clock_seq         |             node              |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
|                              node                             |
+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
```
As a consequence, it might be difficult to find a generic conversion method.